File: java/src/test/java/antipatterns/solution/BlogRefactoredTests.java
package antipatterns.solution;

import blog.Article;
import blog.Comment;
import lombok.val;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.time.LocalDate;
import java.time.temporal.ChronoUnit;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.within;

class BlogRefactoredTests {
    private final String text = "Amazing article !!!";
    private final String author = "Pablo Escobar";
    private Article article;

    @BeforeEach
    void init() {
        article = new Article(
                "Lorem Ipsum",
                "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore"
        );
    }

    @Test
    void it_should_add_a_new_comment_in_the_article_including_given_text_and_author() {
        val result = article.addComment(text, author);

        assertThat(result.isRight()).isTrue();
        assertComment(result.get().getComments().head(), text, author);
    }

    private void assertComment(Comment comment,
                               String expectedText,
                               String expectedAuthor) {
        assertThat(comment.getText()).isEqualTo(expectedText);
        assertThat(comment.getAuthor()).isEqualTo(expectedAuthor);
        assertThat(comment.getCreationDate()).isCloseTo(LocalDate.now(), within(1, ChronoUnit.SECONDS));
    }

    @Test
    void it_should_return_an_error_when_adding_an_existing_comment() {
        val result =
                article.addComment(text, author)
                        .map(a -> a.addComment(text, author))
                        .flatMap(r -> r);

        assertThat(result.isLeft()).isTrue();
        assertThat(result.getLeft())
                .hasSize(1)
                .allMatch(error -> error.getDescription().equals("Comment already in the article"));
    }
}

File: scala/src/test/scala/anti/patterns/solution/BlogRefactoredTests.scala
package anti.patterns.solution

import anti.patterns.solution.BlogRefactoredTests._
import demo.blog.{Article, BlogService, Comment}
import org.scalatest.EitherValues
import org.scalatest.flatspec.AnyFlatSpec

import java.time.LocalDate

class BlogRefactoredTests extends AnyFlatSpec with EitherValues {
  private val blogService = new BlogService()

  it should "add a new comment in an empty Article including given text / author" in {
    val updatedArticle = blogService.addComment(emptyArticle, text, author)
    assert(updatedArticle.isRight)
    assertAddedComment(updatedArticle.value, text, author)
  }

  it should "add a new comment in an Article containing existing comments" in {
    val newText = "Finibus Bonorum et Malorum"
    val newAuthor = "Al Capone"

    val updatedArticle = blogService.addComment(articleWith1Comment, newText, newAuthor)

    assert(updatedArticle.isRight)
    assert(updatedArticle.value.comments.size == 2)
    assertAddedComment(updatedArticle.value, newText, newAuthor)
  }

  it should "return an error when adding existing comment" in {
    val updatedArticle = blogService.addComment(articleWith1Comment, text, author)

    assert(updatedArticle.isLeft)
    assert(updatedArticle.left.value.size == 1)
    assert(updatedArticle.left.value.head.description == "Comment already in the article")
  }
}

object BlogRefactoredTests {
  private val text = "Amazing article !!!"
  private val author = "Pablo Escobar"

  private val emptyArticle: Article = articleWithComments(List.empty[Comment])
  private val articleWith1Comment: Article = articleWithComments(List(Comment(text, author, LocalDate.now())))

  private def articleWithComments(comments: List[Comment]): Article =
    Article(
      "Lorem Ipsum",
      "consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore",
      comments
    )

  private def assertAddedComment(
      article: Article,
      expectedText: String,
      expectedAuthor: String
  ): Unit = {
    val addedComment = article.comments.head
    assert(addedComment.text == expectedText)
    assert(addedComment.author == expectedAuthor)
    assert(addedComment.creationDate.isEqual(LocalDate.now))
  }
}
